File: server/common/src/main/java/su/plo/voice/server/command/VoiceReconnectCommand.java
package su.plo.voice.server.command;

import lombok.RequiredArgsConstructor;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import su.plo.slib.api.chat.component.McTextComponent;
import su.plo.slib.api.command.McCommand;
import su.plo.slib.api.command.McCommandSource;
import su.plo.slib.api.server.entity.player.McServerPlayer;
import su.plo.voice.api.server.player.VoiceServerPlayer;
import su.plo.voice.server.BaseVoiceServer;

@RequiredArgsConstructor
public final class VoiceReconnectCommand implements McCommand {

    private final BaseVoiceServer voiceServer;

    @Override
    public void execute(@NotNull McCommandSource source, @NotNull String[] arguments) {
        if (!(source instanceof McServerPlayer)) {
            source.sendMessage(McTextComponent.translatable("pv.error.player_only_command"));
            return;
        }

        McServerPlayer player = (McServerPlayer) source;
        VoiceServerPlayer voicePlayer = voiceServer.getPlayerManager().getPlayerById(player.getUuid())
                .orElseThrow(() -> new IllegalStateException("how?"));

        source.sendMessage(McTextComponent.translatable("pv.command.reconnect.message"));
        voiceServer.getUdpConnectionManager().removeConnection(voicePlayer);
        voiceServer.getTcpPacketManager().requestPlayerInfo(voicePlayer);
    }

    @Override
    public boolean hasPermission(@NotNull McCommandSource source, @Nullable String[] arguments) {
        return source.hasPermission("pv.reconnect");
    }
}

File: server/common/src/main/kotlin/su/plo/voice/server/command/VoiceReconnectCommand.kt
package su.plo.voice.server.command

import dev.apehum.mcdsl.command.literalCommand
import su.plo.slib.api.chat.component.McTextComponent
import su.plo.slib.api.command.brigadier.McBrigadierSource
import su.plo.slib.api.server.entity.player.McServerPlayer
import su.plo.voice.api.server.connection.TcpServerPacketManager
import su.plo.voice.api.server.connection.UdpServerConnectionManager
import su.plo.voice.api.server.event.connection.UdpClientDisconnectedEvent
import su.plo.voice.api.server.player.VoicePlayerManager
import su.plo.voice.api.server.player.VoiceServerPlayer

fun voiceReconnectCommand(
    playerManager: () -> VoicePlayerManager<VoiceServerPlayer>,
    udpConnectionManager: () -> UdpServerConnectionManager,
    tcpPacketManager: () -> TcpServerPacketManager,
) =
    literalCommand<McBrigadierSource>("vrc") {
        requires { it.source.hasPermission(Permission.RECONNECT) }

        executes {
            val player = source.executor as? McServerPlayer
                ?: run {
                    source.sendFeedback(McTextComponent.translatable("pv.error.player_only_command"))
                    return@executes
                }

            val voicePlayer = playerManager().getPlayerById(player.uuid)
                .orElseThrow { IllegalArgumentException("Player ${player.name} not found") }

            source.sendFeedback(McTextComponent.translatable("pv.command.reconnect.message"))
            udpConnectionManager().removeConnection(voicePlayer, UdpClientDisconnectedEvent.Reason.RECONNECT)
            tcpPacketManager().requestPlayerInfo(voicePlayer)
        }
    }
